x-pack/plugins/proxy/server/proxy.ts

```diff
@@ -0,0 +1,291 @@
+      );
+      const res = h.response('Service Unavailable');
+      res.code(503);
+      return res;
+    }
+
+    url.hostname = node.node;
+    try {
+      const { payload } = await this.wreck.get(url.toString(), { json: true });
+      const res = h.response(payload);
+      res.code(200);
+      res.header('content-type', 'application/json');
+      return res;
+    } catch (err) {
+      this.log.warn(
+        `Unable to complete request to ${node.node} for ${resource} because ${err.message}`
+      );
+      const res = h.response('Internal server error');
+      res.code(500);
+      return res;
+    }
+  }
+
+  public getAllocation(): Observable<[string, RoutingNode]> {
+    return from(this.routingTable);
+  }
+
+  private setTimer() {
+    if (this.timer) return;
+    this.log.debug('Set timer to updateNodeMap');
+    this.timer = setInterval(async () => {
+      this.log.debug('Updating node map');
+      await this.mainLoop();
+    }, this.updateInterval);
+  }
+
+  private updateRoutingTable(routingTable: { [key: string]: RoutingNode }): void {
+    const currentRoutes = [...this.routingTable.keys()];
+    for (const [key, node] of Object.entries(routingTable)) {
+      this.routingTable.set(key, node);
+      const idx = currentRoutes.findIndex(k => k === key);
+      if (idx) currentRoutes.splice(idx, 1);
+    }
+
+    for (const key of currentRoutes.values()) {
+      this.routingTable.delete(key);
+    }
+  }
+
+  private async getNodeList(): Promise<NodeList> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      _source: true,
+    };
+    const data: ClusterDoc = await client.callAsInternalUser('get', params);
+    this.updateRoutingTable(data.routing_table);
+    const nodes: NodeList = new Map(Object.entries(data.nodes));
+    return nodes;
+  }
+
+  private async updateNodeList(nodes: NodeList): Promise<void> {
+    if (!this.elasticsearch) {
+      throw new Error('You must call setup first');
+    }
+    const doc = {
+      nodes: map2obj(nodes),
+      routing_table: map2obj(this.routingTable),
+    };
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      doc,
+    };
+    await client.callAsInternalUser('update', params);
+  }
+
+  private updateLocalNode(nodes: NodeList, finishTime: number): NodeList {
+    nodes.set(this.nodeName, {
+      lastChangedTime: finishTime,
+      lastUpdate: finishTime,
+    });
+    return nodes;
+  }
+
+  private async mainLoop(): Promise<any> {
+    const nodes = await this.getNodeList();
+    const finishTime = new Date().getTime();
+
+    for (const [key, node] of nodes.entries()) {
+      if (!node || finishTime - node.lastUpdate > this.timeoutThreshold) {
```

Comment:
looks like it's important piece I'd extract it at least to a separate method or even create a `Nodes` model